### ISSUE TITLE
Rapid-Fire Artillery slowest with moderate-high Power adjacency

### ISSUE THREAD

**Issue by SwayGin-FAF** [FAForever/fa#1444]
If an Aeon Rapid-Fire artillery has exactly 2 T3 pgens on opposite sides, it will get the speed bonus, but when another T3 is placed on one of the two remaining sides, it actually becomes strictly slower than if it had no power adjacency.

Building T2 power exactly in the center (recall it's 5x5, which may be part of the issue) of each face seemed to be ok, but after surrounding the remaining 2x4 adjacent squares with T1 the problem occurred again, and so it is not solely an issue with T3 Pgens.

When in this state, the "recharge timer" bar will progress rapidly due to the adjacency, as expected, but it will stop at ~ 97% (leaving a small but noticeable gap), after which there will be a giant delay before firing another shot.  When tested side-by-side with another RFA which had nothing around it, the RFA with 3 T3 pgens actually fired slower.  If you remove some of the pgens (e.g., with 3 T3 pgens, reclaim one of them), it will perform correctly again, i.e. a bit faster than with no power adjacency, making it quite simple to test.

This problem does not occur, however, with the Mavor or any factions' T3 heavy artillery.  Also, the direction the artillery was pointed in had no effect, i.e. this happened when it was placed both rotated and normally.

**Comment by aeoncleanse** [comment]
I'm sure we fixed this in the past...

**Comment by SwayGin-FAF** [comment]
No.

Replay 5089450

**Comment by shalkya** [comment]
i thought you said this was no possibility to fix it ?

**Comment by duk3luk3** [comment]
This is an area where @Crotalus should have some insight.

**Comment by aeoncleanse** [comment]
https://github.com/FAForever/fa/issues/350

**Comment by aeoncleanse** [comment]
So yeah, that issue shoulda fixed it. What do guys!

**Comment by aeoncleanse** [comment]
Right, I've confirmed the bug.

None - 31
One T3 - 28
Two T3 - 25
Three T3 - 44 - Bug confirmed. The progress bar advances at the correct rate, then pauses just before the end.

What the hell...

**Comment by aeoncleanse** [comment]
Logging the value AdjRoFMod from this code

```            local val = BuffCalculate(unit, buffName, 'RateOfFire', 1)

            for i = 1, unit:GetWeaponCount() do
                local wep = unit:GetWeapon(i)
                local bp = wep:GetBlueprint()
                -- Set new rate of fire based on blueprint rate of fire
                wep:ChangeRateOfFire(bp.RateOfFire / val)
                wep.AdjRoFMod = val
            end```

Shows that the value goes down by 0.1 per T3 PGen added, from 1, as dictated by AdjacencyBuffs.lua.

It should result in ChangeRateOfFire giving a 22 tick interval, but for some reason it's delayed!

**Comment by Crotalus** [comment]
The weapon waits on the economy event which doesn't take AdjRoFMod into account when calculating what duration the drain should have, in this case the drain is over 2 seconds.

So if Rate of fire gets <2s per shot it isn't ready at the next OnFire and then waits for another round instead. Getting "out of sync" with the rate of fire so to speak.

Either change drain time to 1s so it's lower than maxium weapon rate of fire or change the EconomyEvent code to automatically reduce timespan of drain:

```
...
 local time = nrgReq / nrgDrain
if self.AdjRoFMod then
    time = time * self.AdjRoFMod
end
```

Would result in a bit confusion with the adjacency bonus giving lower energy drain which then is countered by higher energy / s displayed in UI due to energy being drained faster.

**Comment by aeoncleanse** [comment]
Actually, getting rid of the MuzzleSalvoDelay to 0 fixes this too. Does that prevent it watching the economy triggers?

**Comment by Crotalus** [comment]
Probably getting squeezed into ~2s timespan in that case, I think maximum rate of fire here is something like ~0.57 which would be 1.75 + that 0.2 delay + something else maybe so it could make it or break it yeah

**Comment by aeoncleanse** [comment]
Yep, the energy drain rate does the trick here. Thanks for the tip Crotalus.

**Comment by speed2CZ** [comment]
Should be fixed by #1455
